FreeRTOS-Plus/Source/Application-Protocols/network_transport/freertos_plus_tcp/using_wolfSSL/using_wolfSSL.c

```diff
@@ -0,0 +1,409 @@
+    {
+        socketStatus = Sockets_Connect( &( pNetworkContext->tcpSocket ),
+                                        pHostName,
+                                        port,
+                                        receiveTimeoutMs,
+                                        sendTimeoutMs );
+
+        if( socketStatus != 0 )
+        {
+            LogError( ( "Failed to connect to %s with error %d.",
+                        pHostName,
+                        socketStatus ) );
+            returnStatus = TLS_TRANSPORT_CONNECT_FAILURE;
+        }
+    }
+
+    /* Initialize tls. */
+    if( returnStatus == TLS_TRANSPORT_SUCCESS )
+    { 
+        returnStatus = initTLS();
+    }
+
+    /* Perform TLS handshake. */
+    if( returnStatus == TLS_TRANSPORT_SUCCESS )
+    {
+        returnStatus = tlsSetup( pNetworkContext, pHostName, pNetworkCredentials );
+    }
+
+    /* Clean up on failure. */
+    if( returnStatus != TLS_TRANSPORT_SUCCESS )
+    {
+        if( pNetworkContext->tcpSocket != FREERTOS_INVALID_SOCKET )
+        {
+            FreeRTOS_closesocket( pNetworkContext->tcpSocket );
+        }
+    }
+    else
+    {
+        LogInfo( ( "(Network connection %p) Connection to %s established.",
+                   pNetworkContext,
+                   pHostName ) );
+    }
+
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+void TLS_FreeRTOS_Disconnect( NetworkContext_t * pNetworkContext )
+{
+    WOLFSSL* pSsl = pNetworkContext->sslContext.ssl;
+
+    /* shutdown an active TLS connection */
+    wolfSSL_shutdown(pSsl);
+
+    /* cleanup WOLFSSL object */
+    wolfSSL_free(pSsl);
+    pNetworkContext->sslContext.ssl = NULL;
+
+    /* Call socket shutdown function to close connection. */
+    Sockets_Disconnect(pNetworkContext->tcpSocket);
+
+    /* free WOLFSSL_CTX object*/
+    WOLFSSL_CTX* pCtx = pNetworkContext->sslContext.ctx;
+    wolfSSL_CTX_free(pCtx);
+    pNetworkContext->sslContext.ctx = NULL;
+
+    wolfSSL_Cleanup();
+}
+
+/*-----------------------------------------------------------*/
+
+int32_t TLS_FreeRTOS_recv( NetworkContext_t * pNetworkContext,
+                           void * pBuffer,
+                           size_t bytesToRecv )
+{
+    int32_t tlsStatus = 0;
+    int  iResult = 0;
+    WOLFSSL* pSsl = pNetworkContext->sslContext.ssl;
+    
+    iResult = wolfSSL_read( pSsl, pBuffer, bytesToRecv );
+
+    
+    if (iResult > 0)
+        tlsStatus = iResult;
+    else
+        tlsStatus = 0;
+        
+    return tlsStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+int32_t TLS_FreeRTOS_send( NetworkContext_t * pNetworkContext,
+                           const void * pBuffer,
+                           size_t bytesToSend )
+{
+    int32_t tlsStatus = 0;
+    int  iResult = 0;
+    
+    iResult = wolfSSL_write(pNetworkContext->sslContext.ssl, pBuffer, bytesToSend);
+    
+    if (iResult > 0)
```

Comment:
Same here

Reply:
Added wolfSSL_want_write and wolfSSL_state calling to get more detailed error.